### PR TITLE
Returning 403 status code in case of authorization failures

### PR DESCRIPTION
**Pull request by soumitra-st**
Currently, if SQL query execution has authorization issues, the return code is 200 and the errorCode 180 is set in the response JSON. This happens for both controller and broker endpoints.

SQL execution using Controller:
% curl -v -X POST -H "accept: application/json" -H "Content-Type: application/json" -d '{"sql":"select * from transcript limit 10","trace":false,"queryOptions":""}' http://localhost:9000/sql ; echo
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:9000...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> POST /sql HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.88.1
> accept: application/json
> Content-Type: application/json
> Content-Length: 75
>
< HTTP/1.1 200 OK
< Pinot-Controller-Host: soumitras-mbp.attlocal.net
< Pinot-Controller-Version: 1.0.0-sk-bf66c10b03399e8682ed69cf7c087127ca986c36
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS, DELETE
< Access-Control-Allow-Headers: *
< Content-Type: application/json
< Content-Length: 1023
<
* Connection #0 to host localhost left intact
{"requestId":"1262666501000000001","exceptions":[{"message":null,"errorCode":180}],"numServersQueried":0,"numServersResponded":0,"numSegmentsQueried":0,"numSegmentsProcessed":0,"numSegmentsMatched":0,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"numDocsScanned":0,"numEntriesScannedInFilter":0,"numEntriesScannedPostFilter":0,"numGroupsLimitReached":false,"totalDocs":0,"timeUsedMs":0,"offlineThreadCpuTimeNs":0,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":0,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":0,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":0,"realtimeTotalCpuTimeNs":0,"segmentStatistics":[],"traceInfo":{},"numRowsResultSet":0,"minConsumingFreshnessTimeMs":0,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0}

SQL execution using Broker:
% curl -v -X POST -H "accept: application/json" -H "Content-Type: application/json" -d '{"sql":"select * from transcript limit 10","trace":false,"queryOptions":""}' http://localhost:8000/query/sql ; echo
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> POST /query/sql HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.88.1
> accept: application/json
> Content-Type: application/json
> Content-Length: 75
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 1023
<
* Connection #0 to host localhost left intact
{"requestId":"1262666501000000000","exceptions":[{"message":null,"errorCode":180}],"numServersQueried":0,"numServersResponded":0,"numSegmentsQueried":0,"numSegmentsProcessed":0,"numSegmentsMatched":0,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"numDocsScanned":0,"numEntriesScannedInFilter":0,"numEntriesScannedPostFilter":0,"numGroupsLimitReached":false,"totalDocs":0,"timeUsedMs":0,"offlineThreadCpuTimeNs":0,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":0,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":0,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":0,"realtimeTotalCpuTimeNs":0,"segmentStatistics":[],"traceInfo":{},"numRowsResultSet":0,"minConsumingFreshnessTimeMs":0,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0}

This PR changes the return code to 403, sample runs below:
% curl -v -X POST -H "accept: application/json" -H "Content-Type: application/json" -d '{"sql":"select * from transcript limit 10","trace":false,"queryOptions":""}' http://localhost:9000/sql ; echo
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:9000...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> POST /sql HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.88.1
> accept: application/json
> Content-Type: application/json
> Content-Length: 75
>
< HTTP/1.1 403 Forbidden

% curl -v -X POST -H "accept: application/json" -H "Content-Type: application/json" -d '{"sql":"select * from transcript limit 10","trace":false,"queryOptions":""}' http://localhost:8000/query/sql ; echo
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> POST /query/sql HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.88.1
> accept: application/json
> Content-Type: application/json
> Content-Length: 75
>
< HTTP/1.1 403 Forbidden

label is 'bugfix'.

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 1]
```diff
@@ -385,8 +387,7 @@ private BrokerResponseNative handleRequest(long requestId, String query,
       if (!hasTableAccess) {
         _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.REQUEST_DROPPED_DUE_TO_ACCESS_ERROR, 1);
         LOGGER.info("Access denied for request {}: {}, table: {}", requestId, query, tableName);
-        requestContext.setErrorCode(QueryException.ACCESS_DENIED_ERROR_CODE);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:388]: We should still set the request context

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java [hunk 2]
```diff
@@ -390,7 +396,9 @@ public String sendPostRaw(String urlStr, String requestStr, Map<String, String>
       /*if (LOG.isInfoEnabled()){
         LOGGER.info("The http response code is " + responseCode);
       }*/
-      if (responseCode != HttpURLConnection.HTTP_OK) {
+      if (responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:399]: Should we first check response code before getting the output stream?

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java [hunk 3]
```diff
@@ -121,6 +121,8 @@ public void processSqlQueryGet(@ApiParam(value = "Query", required = true) @Quer
       }
       BrokerResponse brokerResponse = executeSqlQuery(requestJson, makeHttpIdentity(requestContext), true);
       asyncResponse.resume(brokerResponse.toJsonString());
+    } catch (WebApplicationException wae) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java:124]: Put some error log? Also do we have a metric tracking the auth failure?
Same for other places

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:388]: Please check all usage of `hasAccess()`, seems there are other places also need to be handled

**Review by Jackie-Jiang — COMMENTED** [review]
Please take a look at the test failure. We want to add/reuse existing test to verify this new behavior

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 1 above)
Comment (soumitra-st) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:388]: Checked and handled other calls of hasAccess.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java (hunk 2 above)
Comment (soumitra-st) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:401]: output stream is used to send the payload, input and error streams are read after checking the status code. The logic looks correct.

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java (hunk 3 above)
Comment (soumitra-st) [reply pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotClientRequest.java:124]: Added log. There is a Meter metric REQUEST_DROPPED_DUE_TO_ACCESS_ERROR("requestsDropped") which is tracking at several places, but let me file a task to review and add at the missing places, such as request filter.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:401]: Ah, sorry I misread the code, nvm

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**